contracts/settling_game/L05_Wonders.cairo

```diff
@@ -0,0 +1,280 @@
+%lang starknet
+%builtins pedersen range_check bitwise
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
+from starkware.cairo.common.math import assert_in_range, assert_nn, assert_nn_le, unsigned_div_rem, assert_not_zero
+from starkware.cairo.common.math_cmp import is_nn_le
+from starkware.cairo.common.hash_state import hash_init, hash_update, HashState
+from starkware.cairo.common.alloc import alloc
+from starkware.starknet.common.syscalls import get_caller_address, get_block_timestamp
+from starkware.cairo.common.uint256 import Uint256, uint256_eq
+
+from contracts.settling_game.utils.general import scale
+from contracts.settling_game.interfaces.imodules import (
+    IModuleController, IS02_Resources, IS01_Settling, IL04_Calculator, IS05_Wonders)
+
+from contracts.settling_game.utils.game_structs import RealmData, ResourceUpgradeIds
+from contracts.settling_game.utils.general import unpack_data
+
+from contracts.token.ERC20.interfaces.IERC20 import IERC20
+from contracts.token.ERC1155.interfaces.IERC1155 import IERC1155
+from contracts.settling_game.interfaces.realms_IERC721 import realms_IERC721
+from contracts.settling_game.interfaces.s_realms_IERC721 import s_realms_IERC721
+
+
+# #### Module L05 ##########
+#                        #
+# Logic for Wonders      #
+#                        #
+##########################
+
+@storage_var
+func controller_address() -> (address : felt):
+end
+
+@constructor
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        address_of_controller : felt):
+    controller_address.write(address_of_controller)
+    return ()
+end
+
+@external
+func pay_wonder_upkeep{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        epoch : felt,
+        token_id : Uint256):
+    alloc_locals
+    let ( controller ) = controller_address.read()
+    let ( caller ) = get_caller_address()
+
+    let ( treasury_address ) = IModuleController.get_treasury_address(contract_address=controller)
+    let ( resources_address ) = IModuleController.get_resources_address(contract_address=controller)
+    let ( wonders_state_address ) = IModuleController.get_module_address(contract_address=controller, module_id=9)
+
+    # calculator logic contract
+    let ( calculator_address ) = IModuleController.get_module_address(
+        contract_address=controller, module_id=7)
+
+    # get currenty epoch
+    let ( current_epoch ) = IL04_Calculator.calculateEpoch(calculator_address)
+
+    assert_nn_le(current_epoch, epoch)
+
+    # Set upkept for epoch
+    IS05_Wonders.set_wonder_epoch_upkeep(wonders_state_address, epoch, token_id, 1)
+
+    # Upkeep cost, to be moved
+    let ( local upkeep_token_ids : Uint256* ) = alloc()
```

Comment:
`safeBatchTransferFrom` [takes](https://github.com/BibliothecaForAdventurers/realms-contracts/blob/wonder-tax/contracts/token/ERC1155/interfaces/IERC1155.cairo#L22) only `felts`, not `Uint256`